### PR TITLE
Refactor of RewriteDataFiles and BinPackStrategy

### PR DESCRIPTION
**Pull request by RussellSpitzer**
Remove explicit staregy api in Rewrite Data Files APi so that we can more easily add strategies with custom arguements, such as "sort(SortOrder ...)

Change BinPack Defaults and organization to better handle small and large files
Removes the output size parameter and replaces it with a constant check clause to see if the output of the rewrite would produce a target size file or the input exceeds the MIN_INPUT_FILES param

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java [hunk 1]
```diff
@@ -45,20 +44,12 @@
 abstract class BinPackStrategy implements RewriteStrategy {
 
   /**
-   * Minimum number of files that need to be in a file group to be considered
-   * for rewriting. This is considered in conjunction with {@link MIN_OUTPUT_FILES}, both
-   * conditions must pass to consider a group of files to be rewritten.
+   * The minimum number of files that need to be in a file group for it to be considered for
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:47]: Looks like we have to update the Javadoc for the class as well.

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java [hunk 2]
```diff
@@ -45,20 +44,12 @@
...
+   * compaction if the total size of that group is not the target size. This can also be thought of as
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:48]: `... if the total file group size is less than the target file size`?

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java [hunk 3]
```diff
@@ -45,20 +44,12 @@
...
+   * the maximum number of non-target-size files that should remain in a file group (parititon) after rewriting.
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:49]: nit: `should` -> `can`?

Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:49]: I couldn't say why but I think it should be "could" ... replaced

Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:49]: Also typo on partition, I see now

**Review by aokolnychyi — APPROVED** [review]

**Comment by aokolnychyi** [comment]
Thanks, @RussellSpitzer!

**Pull request merged**